Create config-in-database.rst

wiersgallak opened the pull request.

jasonblais commented on source/administration/command-line-tools.rst:
```diff
@@ -106,6 +106,7 @@ mattermost
     .. code-block:: none
 
       -c, --config {string}   Configuration file to use. (default "config.json")
+      -- disableconfigwatch {boolean}   When set config.json will not be loaded from disk when the file is changed (default "false")
```
Not sure I understand - is it that if it's disabled, config.json will not be loaded from the disk when the JSON file is updated?

wiersgallak replied:
@lieut-data can you please help me with this question?

lieut-data replied:
First, a minor spacing change:
```suggestion
      --disableconfigwatch {boolean}   When set config.json will not be loaded from disk when the file is changed (default "false")
```

Second, this comes right out of the `mattermost CLI` help, so if we change, we should probably do it in both places. Here's a different stab at the help:
```suggestion
      --disableconfigwatch {boolean}   When set, config.json will not be reloaded automatically when another process changes it (default "false")
```

Downside it's a bit long, and already the longest help item.

jasonblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,26 @@
+Configuration in the Mattermost Database
+=========================================
+A new configuration database was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to allow using the database for the single source of truth of the active configuration of your Mattermost installation. This provides a convenient way to manage configuration changes from the system console and provides a failsafe in high availability deployments ...
```
Do you mean `A new configuration option`

jasonblais replied:
- `system console` -> `System Console`
- `failsafe` may not be obvious for non-English native spakers

jasonblais replied:
Wondering if there's a simpler way to write the last sentence. Or maybe having the benefits in bullets helps with readability

wiersgallak replied:
I have addressed this feedback.

jasonblais replied:
There's one more `system console` -> `System Console` 🙏

wiersgallak replied:
Thanks - missed that one.

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+
+Benefits to using this option include: 
+
+  - Conveniently manage configuration changes directly from the System Console.
```
```suggestion
  - Conveniently manage configuration changes directly from the System Console, even in High Availability deployments and read-only containerized environments.
```

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+Benefits to using this option include: 
+
+  - Conveniently manage configuration changes directly from the System Console.
+  - In high availability deployments, configuration changes are applied to all nodes instantaneously as possible removing the need to update individual configuration files on each server and ensures all servers have the correct configurations in the event  a server in the node goes down.
```
```suggestion
  - Ensure all servers in a High Availability deployment have the same configuration, even when new servers are added to the cluster.
  - Automatically deploy SAML certificates and keys to all servers in the cluster.
```

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+
+  - Conveniently manage configuration changes directly from the System Console.
+  - In high availability deployments, configuration changes are applied to all nodes instantaneously as possible removing the need to update individual configuration files on each server and ensures all servers have the correct configurations in the event  a server in the node goes down.
+  - Also in high availability deployments, the read-only state of the System Console is removed.   
```
Strictly speaking, I'd say this isn't a "benefit".

```suggestion
```

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+  - In high availability deployments, configuration changes are applied to all nodes instantaneously as possible removing the need to update individual configuration files on each server and ensures all servers have the correct configurations in the event  a server in the node goes down.
+  - Also in high availability deployments, the read-only state of the System Console is removed.   
+
+To change your server from reading the config.json file to using the configuration in database: 
```
Can't make these suggestions all in place, but propose replacing as follows:

```
To start using configuration in database, pass the database connection string via the `--config` flag or `MM_CONFIG` environment variable. For example:

    ```
    ./mattermost --config="postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable\u0026connect_timeout=10"
    ```

To migrate an existing config.json into the database, use the ``config migrate`` `command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost-config-migrate>`_. For example:

    ```
    ./mattermost config migrate  path/to/config.json "postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable&connect_timeout=10"
    ```

Any existing SAML certificates and private keys will also be migrated to the database.
```

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+  
+2. Set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true to disable automatically watching this default config.json and effecting changes.
+
+Once you are using the configuration in the database, changes to configuration settings will create a new entry in the configuration database, storing a copy of the previous configuration as a json blob. 
```
```suggestion
With configuration in the database enabled, any changes to the configuration will recorded the new, full configuration as a new row in the `Configurations` table. Note that `ClusterSettings.ReadOnlyConfig` is ignored when using configuration in the database, enabling full use of the system console.
```

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+Any environment variable configurations will continue to override settings from config.json, triggering them to appear read-only in the system console. When reading the configuration in the database, ClusterSettings.ReadOnlyConfig setting is automatically disabled as the database configuration ignores the config read only flag. 
+
+.. note::
+    If using SAML, ensure the SAML certificates and keys are accessible to also migrate into the database.
```
Consolidated into the above.

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,32 @@
+
+Once you are using the configuration in the database, changes to configuration settings will create a new entry in the configuration database, storing a copy of the previous configuration as a json blob. 
+
+Any environment variable configurations will continue to override settings from config.json, triggering them to appear read-only in the system console. When reading the configuration in the database, ClusterSettings.ReadOnlyConfig setting is automatically disabled as the database configuration ignores the config read only flag. 
```
```suggestion
Note that environment variable overrides remain fully supported, and will no longer be written back to the configuration when a different change is saved.
```

lieut-data commented on source/administration/config-settings.rst:
```diff
@@ -8,6 +8,9 @@ Mattermost configuration settings are maintained in the configuration file ``con
 The default location of ``config.json`` is in the ``mattermost/config`` directory. Mattermost must have write permissions to ``config.json``, otherwise changes made in the System Console will have no effect.
 
+**Configuration in Database**
+Storing configuration in the database is supported as experimental in v5.10 and later.  Please see for more information on how to set this up `here <https://docs.mattermost.com/administration/config-in-database.html>`_.  
```
Do we need to classify this as experimental at this point?

wiersgallak commented:
@lieut-data & @jasonblais Thank you for your review!  I have incorporated most of the changes and suggestions you requested.  

I did not add anything on reverting the migration, as this is how we want most future installations to do their configuration. I added the recommendation for HA environments, but looking for feedback if we want to be more exclusive in our recommendations (maybe adding what scenarios keeping config.json would be more appropriate ? ie: high level of customization of settings? 0/5)

jasonblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+=========================================
+A new configuration option was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database for the single source of truth of the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration tabl ...
+
+Mattermost has been running our community server on this option since the feature was released and recommends using it to those in High Availability deployments.  
```
```suggestion
Mattermost has been running our `community server <https://community.mattermost.com>`_ on this option since the feature was released, and recommends using it to those in :doc:`High Availability deployments <cluster>`.
```

jasonblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+Mattermost has been running our community server on this option since the feature was released and recommends using it to those in High Availability deployments.  
+
+Benefits to using this option include: 
```
```suggestion
Benefits to using this option:
```

jasonblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+  - Ensure all servers in a High Availability deployment have the same configuration, even when new servers are added to the cluster.
+  - Automatically deploy SAML certificates and keys to all servers in the cluster.
+
+To start using configuration in database, pass the database connection string via the `--config` flag or `MM_CONFIG` environment variable. For example:
```
```suggestion
To start using configuration in database, pass the database connection string via the ``--config`` flag or ``MM_CONFIG`` environment variable. For example:
```

jasonblais commented on source/administration/config-in-database.rst:
````diff
@@ -0,0 +1,39 @@
+
+To migrate an existing config.json into the database, use the ``config migrate`` `command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost-config-migrate>`_. For example:
+
+    ```
````
This needs the code block formatting like

```
  .. code-block:: 

    ./mattermost config migrate  path/to/config.json "postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable&connect_timeout=10"
```

jasonblais replied:
(Same for the code block above this one)

jasonblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+Any existing SAML certificates and private keys will also be migrated to the database.
+
+Addtitionally, to disable automatically watching the default config.json and effecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.
```
```suggestion
Addtitionally, to disable automatically watching the default config.json and affecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.
```

jasonblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+Addtitionally, to disable automatically watching the default config.json and effecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.
+
+With configuration in the database enabled, any changes to the configuration are recorded to the `Configurations` table. Furthermore, `ClusterSettings.ReadOnlyConfig` is ignored, enabling full use of the System Console.
```
```suggestion
With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` table. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.
```

wiersgallak commented:
@jasonblais Thanks for additional feedback:
1) Yes that would be the plan, possibly in 6.0. I have updated the config setting docs, moving this setting the config.json only section since it was removed from the System Console UI.
2) Updated the note @lieut-data can you please review this section to ensure accuracy? 
3) Thanks! Updated to include the containerized environments.

amyblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+Configuration in the Mattermost Database
+=========================================
+A new configuration option was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database for the single source of truth of the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration tabl ...
```
```suggestion
A new configuration option was added in `5.10 release <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database as the single source of truth for the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration table in the database. 
```

amyblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+=========================================
+A new configuration option was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database for the single source of truth of the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration tabl ...
+
+Mattermost has been running our `community server <https://community.mattermost.com>`_ on this option since the feature was released, and recommends using it to those in :doc:`High Availability deployments <cluster>`.
```
```suggestion
Mattermost has been running our `community server <https://community.mattermost.com>`_ on this option since the feature was released, and recommends its use for those on :doc:`High Availability deployments <cluster>`.
```

amyblais commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` table. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.
+
+Note that environment variable overrides remain fully supported using the configuration in the database, and will no longer be written back to the configuration when a change is saved, but will be written to the database instead.
```
```suggestion
Note that environment variable overrides remain fully supported using the configuration in the database. They will no longer be written back to the configuration when a change is saved, but will be written to the database instead.
```

amyblais commented on source/administration/config-settings.rst:
```diff
@@ -8,6 +8,9 @@ Mattermost configuration settings are maintained in the configuration file ``con
 The default location of ``config.json`` is in the ``mattermost/config`` directory. Mattermost must have write permissions to ``config.json``, otherwise changes made in the System Console will have no effect.
 
+**Configuration in Database**
+Storing configuration in the database is supported in v5.10 and later.  Please see for more information on how to set this up `here <https://docs.mattermost.com/administration/config-in-database.html>`_.  
```
```suggestion
Storing configuration in the database is supported in v5.10 and later.  Please see more information on how to set this up `here <https://docs.mattermost.com/administration/config-in-database.html>`_.  
```

amyblais commented on source/administration/config-settings.rst:
```diff
@@ -8,6 +8,9 @@ Mattermost configuration settings are maintained in the configuration file ``con
 
 The default location of ``config.json`` is in the ``mattermost/config`` directory. Mattermost must have write permissions to ``config.json``, otherwise changes made in the System Console will have no effect.
 
+**Configuration in Database**
```
This section will be probably need to be added here too: https://docs.mattermost.com/administration/prev-config-settings.html.

wiersgallak replied:
added - thanks for the suggestion

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+Any existing SAML certificates and private keys will also be migrated to the database.
+
+Addtitionally, to disable automatically watching the default config.json and affecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.
```
I don't think this paragraph belongs in this document -- the focus is on switching /to/ database in the config, and this flag does nothing in that configuration.
```suggestion
```

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+Addtitionally, to disable automatically watching the default config.json and affecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.
+
+With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` table. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.
```
With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` and ``ConfigurationFiles`` tables. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.

lieut-data commented on source/administration/config-in-database.rst:
```diff
@@ -0,0 +1,39 @@
+
+With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` table. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.
+
+Note that environment variable overrides remain fully supported using the configuration in the database. They will no longer be written back to the configuration when a change is saved, but will be written to the database instead.
```
In fact, they won't be written anywhere:
```
Note that environment variable overrides remain fully supported using the configuration in the database, allowing per-server customization as required. A long standing bug has also been fixed, ensuring environment variables are no longer written back to the configuration when other changes are made.
```

lieut-data commented on source/administration/config-settings.rst:
```diff
@@ -3515,6 +3506,18 @@ This setting defines the frequency of cluster request time logging for :doc:`../
+
+**False**: Changes made to settings in the System Console are written to ``config.json``.
+
+When running in production it is recommended to set this to true.
```
This statement is no longer true -- recommend removing.
```suggestion
```

lieut-data commented on source/administration/config-settings.rst:
```diff
@@ -3515,6 +3506,18 @@ This setting defines the frequency of cluster request time logging for :doc:`../
+^^^^^^^^^^^^^^^^
+**True**: Changes made to settings in the System Console are ignored.
+
+**False**: Changes made to settings in the System Console are written to ``config.json``.
```
The reference here to `config.json` is the kind of change we may want to holistically update in the future to decouple ourselves from the file-based orientation.

wiersgallak replied:
Makes sense and there are likely many more throughout the docs.  Propose tackling that effort separately.

amyblais commented:
@wiersgallak One thing that was forgotten was to add this in the table of contents.

wiersgallak commented:
Good catch - will put in a quick PR.
